specs/0007-pull-requests.md

```diff
@@ -0,0 +1,144 @@
+# What makes a good pull request?
+
+This document discusses our standards for pull requests and commits.
+
+## Problem Description
+
+A poorly constructed pull request can be a frustrating experience both
+for the person making the pull request and for those people that are
+trying to review it. By setting forth some guidelines in this
+document, we hope to make the review process a useful and productive
+experience for everyone involved.
+
+## Policy
+
+### What makes a good pull request?
+
+A good pull request is **clean**. In most cases it should be made
+against the current version of the primary branch in the repository
+(usually `master`, `main`, or `trunk`). It should be free of syntax
+errors and linting problems. Your local development environment should
+be configured to notify you of those problems in your editor and/or in
+a `pre-commit` hook. Your pull request should pass any unit tests that
+exist in the repository (and if you are introducing a new feature,
+your pull request should include tests appropriate to that feature).
+
```

Comment:
Somewhere we need to mention:
1) More specific lint requirements
2) Which specific pretty printer (ie 'black') should be used and its specific configuration if it is not the default
3) Potentially a project that has just a skeleton that gives the basic support for the tools that they are expected to use.

Reply:
> Somewhere we need to mention:

I don't think these are appropriate to this document, because they're all going to vary by project. We have things written in Python, Go, Shell, YAML, JSON, Markdown, etc, and each language is going to have a different set of tools.

Additionally, it's not my goal to be prescriptive here: different projects may have different requirements (for example, a project that lives within the OpenStack ecosystem may use a specific set of tooling that's not necessary or desirable in another project).

I think we want people to have flexibility in their tooling as long as they meet the requirement of submitting changes that are free of simple errors.

Reply:
I've added a node to the document about the utility of code formatting tools.